### DISCUSSION TITLE
How do I disable hover styles on mobile?

### DISCUSSION THREAD

**Discussion by rightaway** [tailwindlabs/tailwindcss#1739] [Help]
If I create a custom media query that uses https://caniuse.com/#feat=css-media-interaction to target devices that are not mobile, I can do something like `notmobile:hover:font-bold` to make sure the bold style only takes effect not on mobile devices.

The problem is that I'd have to remember to always use `notmobile:hover:` everywhere I'm using `hover:`, which clutters the html and could lead to forgetting to use it in certain places.

A feature where we could set `:hover` to only apply for a certain media query that applies globally would be useful. Often the hover styles cause issues on mobile browsers where they're better off not being there.

Alternatively is there a way to write a media query that automatically appends another style to each style? So that when I create the `notmobile` media query I can tell it to generate all its styles appending `:hover` to it. Then I only need to use `notmobile:font-bold` instead of `notmobile:hover:font-bold`. This feature could be a better way of doing it than what I suggested before because this is more generic and could allow for interesting possibilities.

**Comment by estevanmaito** [comment]
Could you expand on this? Because `:hover` is not even triggered on mobile.

> Often the hover styles cause issues on mobile browsers where they're better off not being there.

It could be the case when you have content that is hidden/revealed on hover, but then thats also a problem for eveybody that's not using a mouse.

> Alternatively is there a way to write a media query that automatically appends another style to each style?

It can be done with [plugins](https://tailwindcss.com/docs/plugins/#app)

**Comment by rightaway** [comment]
`:hover` is triggered on mobile.

**Comment by stefanfisk** [comment]
The behavior of `:hover` is browsers dependent on mobile. Safari for example sets hover on touch, and only removes it once another element is touched.

**Comment by rightaway** [comment]
Exactly and this inconsistency is another reason to want to disable them.

**Comment by VaibhavAcharya** [comment]
Best way is to use media query to detect if a device supports hover and only disable it if it does not.

This can be done by adding this to your `tailwind.config.js` file :
```
module.exports = {
    theme: {
        extend: {
            screens: {
                'betterhover': {'raw': '(hover: hover)'},
            }
        }
    }
}
```

Now you can use it like this : `betterhover:font-bold`

**Reply by CavalcanteLeo** [reply]
This is exactly what im looking for. Best solution, limiting it to a small screen could be falsy mobile

Instead, let the browser decide if hover is supported

**Reply by camin-mccluskey** [reply]
~~This resulted in me getting a constant hover state for all elements on larger screens. Any ideas why that might be? Works perfectly on mobile.~~

~~I'm using Chrome on desktop and iOS Safari~~

Closed the tab and reopened and it appears to be working. No idea what the issue was but thanks for your answer - very useful

**Reply by timmeade** [reply]
Having issues with this.  It does as described for mobile, but now on desktop they are hover all the time.   So all buttons changed to hover state.

**Reply by timmeade** [reply]
OK i misunderstood use:  

you have to:  betterhover:hover:bg-green-100

Now it works fine :-)

**Reply by hyvyys** [reply]
Doesn't work for me in Tailwind 5.0.0-2

**Reply by rohit-k-singh** [reply]
This answer will override the tailwinds min-* and max-* stylings for dynamic responsive screens.

**Reply by illiaChaban** [reply]
@Rksingh620 is right, it overrides arbitrary media queries for me to. For example `min-[320px]:text-center` stops working. Why is that? Is that a bug

**Reply by illiaChaban** [reply]
found the answer. Here's the [issue](https://github.com/tailwindlabs/tailwindcss/issues/13022) and here's the [answer](https://github.com/tailwindlabs/tailwindcss/pull/9558) (it's not supported). There's a workaround tho if you use "addVariant" like [here](https://play.tailwindcss.com/m29hlgZJSc?file=config)

**Reply by brunobely** [reply]
@illiaChaban nice workaround - to use it just `comment-me-out:hover:{classname}`

**Comment by royvanv** [comment]
The downside to @VaibhavAcharya's solution is that it is currently not possible to make it responsive.

**Reply by imjohnlouie04** [reply]
I think it can be done using JIT like for example `md:hover:font-bold`

**Comment by JuanmaMenendez** [comment]
The Tailwind classes:

pointer-events-none
pointer-events-auto

Are everything you need

**Reply by stefanfisk** [reply]
could you provide an example on how to use them to solve this use case?

**Reply by ajmnz** [reply]
> could you provide an example on how to use them to solve this use case?

`<div className="pointer-events-none md:pointer-events-auto"></div>`

**Reply by stefanfisk** [reply]
wouldn't that also disable click?

**Reply by ajmnz** [reply]
> wouldn't that also disable click?

Yes, only use this if that element isn't meant to be clicked. Otherwise, I'd use VaibhavAcharya's solution

**Reply by brunobely** [reply]
Small screen doesn't necessarily mean mobile/touch browser. The only acceptable solutions should use true `@media (hover: hover)` queries.

**Reply by Tofandel** [reply]
If you have hover state in 99% of the cases you want your element to be clickable

**Comment by vdzk** [comment]
@VaibhavAcharya's solution still required me to write out `betterhover:hover:font-bold` which is what @rightaway wanted to avoid in the first place.
[Tailwindcss-touch](https://github.com/steadfast-collective/tailwindcss-touch) allows me to write `hover-hover:font-bold`.

**Reply by vdzk** [reply]
Never mind. Tailwindcss-touch has its own issues. `betterhover:hover:font-bold` it is :(

**Reply by vdzk** [reply]
Now, I found out that this approach only works for me during Chrome's remote debugging. When I unplug the USB cable from my phone, the sticky hover state is back!

**Reply by vdzk** [reply]
Just checked with media query tester. When in remote debugging mode `hover: none`. When in normal mode `hover: hover`. Yep, not only mobile browser lies to me, the remote debugger covers up that lie. Wonderful. On the plus side, tailwind works as expected.

**Reply by kalnode** [reply]
@vdzk So.... **betterhover:font-bold** is still the best thing? Or is it **betterhover:hover:font-bold** as per your earlier comment?

**Reply by jcobert** [reply]
With this, I'm finding that if I want to style the active state, I now have to write redundantly something like `active:font-bold betterhover:active:font-bold` otherwise active styling only applies to touch-enabled devices. Any way around this?

**Comment by ounissi-zakaria** [comment]
If you are using Tailwind with postcss maybe this plugin can help  [postcss-hover-media](https://www.npmjs.com/package/postcss-hover-media-feature).

**Reply by SilencerWeb** [reply]
NOTE: Be careful with this plugin since as I understand it disables hover on touch devices everywhere with no ability to turn it on somewhere you need. E.g. you have a dropdown menu that is opening hover, it won't open on touch devices ¯\\\_(ツ)_/¯

**Comment by egegungordu** [comment]
For anyone bumping into this thread using v3, there's now a tailwind flag you can use to disable hover on mobile devices without downloading anything. [from this pull request](https://github.com/tailwindlabs/tailwindcss/pull/8394)

Just add: 
```
// tailwind.config.js
module.exports = {
  future: {
    hoverOnlyWhenSupported: true,
  },
  // ...
}
```

**Reply by SilencerWeb** [reply]
@bolotskydev @andrewgolovanov

**Reply by SMEETT** [reply]
> For anyone bumping into this thread using v3, there's now a tailwind flag you can use to disable hover on mobile devices without downloading anything. [from this pull request](https://github.com/tailwindlabs/tailwindcss/pull/8394)
> 
> Just add:
> 
> ```
> // tailwind.config.js
> module.exports = {
>   future: {
>     hoverOnlyWhenSupported: true,
>   },
>   // ...
> }
> ```

So far this seems to work fine. Thanks mate!

**Reply by aamir747** [reply]
This works, thanks!

**Reply by CavalcanteLeo** [reply]
some hover are still needed in mobile, eg: dropdown, multi level menu, etc..

So careful when disable it all

**Reply by GeraAlcantara** [reply]
> some hover are still needed in mobile, eg: dropdown, multi level menu, etc..
> 
> So careful when disable it all

But how still be need it if the effect is show in mobile devices unless is touch ? 

I think the future 
`// tailwind.config.js
module.exports = {
  future: {
    hoverOnlyWhenSupported: true,
  },
  // ...
}`

What it does is only wrap the hover in a media query to check if device support hover
`@media (hover: hover)`
Can you elaborate more ?

**Reply by CavalcanteLeo** [reply]
> > some hover are still needed in mobile, eg: dropdown, multi level menu, etc..
> > So careful when disable it all
> 
> But how still be need it if the effect is show in mobile devices unless is touch ?
> 
> I think the future `// tailwind.config.js module.exports = { future: { hoverOnlyWhenSupported: true, }, // ... }`
> 
> What it does is only wrap the hover in a media query to check if device support hover `@media (hover: hover)` Can you elaborate more ?

dropdown can use hover to display the menu, even on mobile:

https://tailwindcomponents.com/component/nestable-dropdown-menu

**Reply by GeraAlcantara** [reply]
Ahh I see in that case that ui component is base on the hover and in mobile devices the user must click to have the drop down, got it

**Reply by sanjayiothakur** [reply]
This works extremely well, thanks for adding this!

**Reply by clintongreen** [reply]
This is amazing, easy fix. Thanks

**Reply by myxomatozis** [reply]
Thanks for this, thrown away dozen of spikes since adding this.

**Reply by CavalcanteLeo** [reply]
⚠️ ⚠️ ⚠️ **DO NOT USE IT** ⚠️ ⚠️ ⚠️ 

**IT CAN LEAD TO VARIOUS BUGS, DO NOT DISABLE HOVER ON MOBILE**

**Reply by tresorama** [reply]
> ⚠️ ⚠️ ⚠️ **DO NOT USE IT** ⚠️ ⚠️ ⚠️
> 
> **IT CAN LEAD TO VARIOUS BUGS, DO NOT DISABLE HOVER ON MOBILE**

Are you referring to disabling it in general or only with this approach taken by tailwind ?

**Reply by adamwathan** [reply]
The only way this can lead to bugs on mobile is if you build your site to depend on hover styles for important interactions, like opening popups on hover instead of click. Don't do that. You can't hover on mobile, so don't rely on hover existing for your site to work.

**Reply by kylemorena** [reply]
Sorry @adamwathan, Can I find it in Tailwind docs?

**Reply by roydigerhund** [reply]
@adamwathan do you think it would make sense to have an option to automatically replace `hover` state with `active` state on touch devices? This way the user would still see the effect while pressing the element but it does not stay in the hover/active state when the press is released.

**Reply by wweiyang** [reply]
Been trying a few other suggested ways that didn't work for me, but this works great, thanks.

**Reply by mikeE-finsight** [reply]
This being the default causes visual bugs because the selector specificity changes between v3 and v4. Here's a reproduction of the issue it causes:

https://play.tailwindcss.com/IvGCGWaLs7

If you hover the button in v3, the specificity of the invalid state is higher but in v4 you have to explicitly set the invalid hover state as well which is pretty unintuitive and clunky

**Reply by Tofandel** [reply]
Is there a way to target some tailwind classes when hover is not supported? Something like `no-hover:block`?

**Reply by rChaoz** [reply]
@Tofandel See my answer below

**Reply by subproject22** [reply]
> @Tofandel See my answer below

Mind update it for v4?

**Reply by rChaoz** [reply]
I will update it these days when I have time

**Reply by rChaoz** [reply]
@subproject22 I have updated [my answer](https://github.com/tailwindlabs/tailwindcss/discussions/1739?sort=top#discussioncomment-9914554) for v4

**Comment by rChaoz** [comment]
**Answer updated for v4** - see the v3 version at the bottom of this post.

Here's my current solution which covers all use-cases, which adds a few custom variants:
- `has-hover:` / `no-hover:` - simply checks for hover support
- the standard `hover:` is [disabled on mobile by default on v4](https://tailwindcss.com/docs/upgrade-guide#hover-styles-on-mobile). This rule is applied on hover on desktop, and never on mobile.
- `hover-always:` - applies the rule on hover on PC, and always on mobile, this rule can also be used with `group-` / `peer-`. For example, can be used for buttons that show on hover.
- `mobile-hover` - replicates the old v3 behaviour, applying the rule on hover on both desktop and mobile (on touch), this can be used for example for button hover animations. Can be used with `group-` / `peer-`.  Do not use this when the rule applying is crucial, e.g. displaying an element, as it's not guaranteed to trigger on mobile.

**Note:** _By PC I mean device supporting hover, and by mobile I mean device not supporting hover, for simplicity._

To use, simply add this to your css file:

```css
@custom-variant has-hover (@media (hover: hover));
@custom-variant no-hover (@media not all and (hover: hover));
@custom-variant hover-always {
  @media (hover: hover) {
    &:hover {
      @slot;
    }
  }
  @media not all and (hover: hover) {
    & {
      @slot;
    }
  }
}
@custom-variant mobile-hover (&:hover);
```

**[ [Playground](https://play.tailwindcss.com/x7OLC2FcLY) | [Cross-test on mobile from desktop](https://www.browserling.com/browse/android15/chrome126/https://play.tailwindcss.com/x7OLC2FcLY) ]**

Sample usage:
```html
<!-- Button is always visible on mobile, otherwise shown when the parent card is hovered -->
<div class="card group">
  ...
  <button class="opacity-0 group-hover-always:opacity-100">...</button>
</div>
```

<details>
<summary>Answer for v3</summary>

In v3, the standard `hover:` does _not_ check for hover support using media queries, so hover states can be activated on mobile as most browsers also emulate hover on touch. You can use `hover:` where this behaviour is acceptable, for example in button hover animations. Do note that in mobile browsers the hover state may not apply on touch, so this should not be used where it's critical, e.g. to show elements.

The config below (which should be added to `tailwind.config.ts`), adds a few custom variants:
- `has-hover:` / `no-hover:` - simply checks for hover support
- `hover-never:` - applies the rule on hover on PC, never on mobile (also provides `group-hover-never` and `peer-hover-never`), this is the same as the updated default `hover:` on v4
- `hover-always:` - applies the rule on hover on PC, always on mobile (also provides `group-hover-always` and `peer-hover-always`)

```ts
import type { Config } from "tailwindcss"
import plugin from "tailwindcss/plugin"

export default {
    // ...
    plugins: [
        plugin(({ addVariant, addUtilities, matchVariant }) => {
            // Hover media queries
            addVariant("has-hover", "@media (hover: hover) and (pointer: fine)")
            addVariant("no-hover", "@media not all and (hover: hover) and (pointer: fine)")

            // Applied on hover if supported, never applied otherwise
            addVariant("hover-never", "@media (hover: hover) and (pointer: fine) { &:hover }")
            matchVariant(
                "group-hover-never",
                (_, { modifier }) => `@media (hover: hover) and (pointer: fine) { :merge(.group${modifier ? "\\/" + modifier : ""}):hover & }`,
                { values: { DEFAULT: "" } },
            )
            matchVariant(
                "peer-hover-never",
                (_, { modifier }) => `@media (hover: hover) and (pointer: fine) { :merge(.peer${modifier ? "\\/" + modifier : ""}):hover & }`,
                { values: { DEFAULT: "" } },
            )

            // Applied on hover if supported, always applied otherwise
            addVariant("hover-always", [
                "@media (hover: hover) and (pointer: fine) { &:hover }",
                "@media not all and (hover: hover) and (pointer: fine)",
            ])
            matchVariant(
                "group-hover-always",
                (_, { modifier }) => [
                    `@media (hover: hover) and (pointer: fine) { :merge(.group${modifier ? "\\/" + modifier : ""}):hover & }`,
                    "@media not all and (hover: hover) and (pointer: fine)",
                ],
                { values: { DEFAULT: "" } },
            )
            matchVariant(
                "peer-hover-always",
                (_, { modifier }) => [
                    `@media (hover: hover) and (pointer: fine) { :merge(.peer${modifier ? "\\/" + modifier : ""}):hover & }`,
                    "@media not all and (hover: hover) and (pointer: fine)",
                ],
                { values: { DEFAULT: "" } },
            )
        }),
    ],
    // ...
} satisfies Config
```

</details>

**Reply by martiuh** [reply]
Thank you very much!

**Reply by delilahw** [reply]
For anyone wanting a quick solution on Tailwind v4, you can use this (e.g. at the top of your `global.css`):
```css
@import "tailwindcss";

/* Replace `hover-never` with `hover` to override the normal `hover` variant,
   which doesn't include the `pointer: fine` constraint in v4. */
@custom-variant hover-never {
  @media (hover: hover) and (pointer: fine) {
    &:hover {
      @slot;
    }
  }
}
```

Usage:
```html
Depending on whether you replaced the original hover variant:
<a class="hover-never:underline"/>
<a class="hover:underline"/>
```

**Reply by Tofandel** [reply]
It is funny that in v4 they disabled hover styles on mobile, but didn't add at least one of those variants by default, which are pretty much necessary to be able to work with hover displaying stuff on desktop but always visible on mobile

**Reply by TalkativeGuy** [reply]
~~Thank you so much for the code!~~
~~Unfortunately, after adding it to my config, an error occured and the app doesn't run:~~
<details><summary>Error</summary>
<p>
```
Uncaught TypeError: os.platform is not a function
    at node_modules/fast-glob/out/utils/path.js (tailwindcss.js?v=3be06300:21719:34)
    at __require2 (chunk-PLDDJCW6.js?v=3be06300:17:50)
    at node_modules/fast-glob/out/utils/index.js (tailwindcss.js?v=3be06300:24858:16)
    at __require2 (chunk-PLDDJCW6.js?v=3be06300:17:50)
    at node_modules/fast-glob/out/managers/tasks.js (tailwindcss.js?v=3be06300:24873:17)
    at __require2 (chunk-PLDDJCW6.js?v=3be06300:17:50)
    at node_modules/fast-glob/out/index.js (tailwindcss.js?v=3be06300:27145:23)
    at __require2 (chunk-PLDDJCW6.js?v=3be06300:17:50)
    at node_modules/tailwindcss/lib/lib/content.js (tailwindcss.js?v=3be06300:27387:46)
    at __require2 (chunk-PLDDJCW6.js?v=3be06300:17:50)
node_modules/fast-glob/out/utils/path.js @ tailwindcss.js?v=3be06300:21719
__require2 @ chunk-PLDDJCW6.js?v=3be06300:17
node_modules/fast-glob/out/utils/index.js @ tailwindcss.js?v=3be06300:24858
__require2 @ chunk-PLDDJCW6.js?v=3be06300:17
node_modules/fast-glob/out/managers/tasks.js @ tailwindcss.js?v=3be06300:24873
__require2 @ chunk-PLDDJCW6.js?v=3be06300:17
node_modules/fast-glob/out/index.js @ tailwindcss.js?v=3be06300:27145
__require2 @ chunk-PLDDJCW6.js?v=3be06300:17
node_modules/tailwindcss/lib/lib/content.js @ tailwindcss.js?v=3be06300:27387
__require2 @ chunk-PLDDJCW6.js?v=3be06300:17
node_modules/tailwindcss/lib/lib/setupTrackingContext.js @ tailwindcss.js?v=3be06300:79703
__require2 @ chunk-PLDDJCW6.js?v=3be06300:17
node_modules/tailwindcss/lib/plugin.js @ tailwindcss.js?v=3be06300:82153
__require2 @ chunk-PLDDJCW6.js?v=3be06300:17
node_modules/tailwindcss/lib/index.js @ tailwindcss.js?v=3be06300:82199
__require2 @ chunk-PLDDJCW6.js?v=3be06300:17
(anonymous) @ tailwindcss.js?v=3be06300:82202
```
</p>
</details>

~~I'm using Tailwind `^3.4.17` and `@nuxtjs/tailwindcss ^6.14.0` Does anyone have any idea what it could mean? I couldn't find any relevant info on the internet~~

Nevermind I imported `import plugin from "tailwindcss"` instead of `import plugin from "tailwindcss/plugin"` because IntelliSence told me that I could
Let's leave it here so nobody repeats my mistake :D
